Source/Dafny/Verifier/Translator.cs

```diff
@@ -1384,7 +1384,7 @@ internal class PredefinedDecls {
         }
         Bpl.Variable resType = new Bpl.Formal(ctor.tok, new Bpl.TypedIdent(ctor.tok, Bpl.TypedIdent.NoName, predef.DatatypeType), false);
         Bpl.Function fn;
-        if (dt is TupleTypeDecl ttd && ttd.Dims == 2) {
+        if (dt is TupleTypeDecl ttd && ttd.Dims == 2 && ttd.NonGhostDims == 2) {
```

Comment:
Should `ttd.Dims == 2` still be in here?

Reply:
Tuples like `(int, int, ghost int)` and `(int, int)` should be translated to different types because we want Boogie to know that instances of the former cannot alias instances of the latter.

Reply:
I'm actually not sure how to observe a difference in the verification time of Dafny. It's a suggestion that I got from Rustan.

Reply:
The following assertion should be provable by the verifier, which requires that the verifier treat `(int, int, ghost int)` as a different type than `(int, int)`.

```
method M(a: array<(int, int, ghost int)>, b: array<(int, int)>)
{
  assert a as object != b;
}
```

Reply:
I'm going to add that example, thanks.

Regarding the if condition on `tdd`, we want to only match tuples of size 2 that have no ghost components, because in that case we want to use the `Tuple2` definition provided by `DafnyPrelude.bpl`. Other conditions (`ttd.NonGhostDims == 2` only, `ttd.Dims == 2` only, `ttd.Dims == 2 && ttd.NonGhostDims == 0`) lead to Boogie errors like "Error: use of undeclared function: Tclass._System.Tuple3OOG".